node/src/neighborhood/node_location.rs

```diff
@@ -0,0 +1,208 @@
+        self.country_code == other.country_code
+    }
+}
+
+impl Eq for NodeLocation {}
+
+pub fn get_node_location(ip: Option<IpAddr>) -> Option<NodeLocation> {
+    match ip {
+        Some(ip_addr) => {
+            let country = CountryCodeFinder::find_country(&COUNTRY_CODE_FINDER, ip_addr);
+            match country {
+                Some(country) => Some(NodeLocation {
+                    country_code: country.iso3166.to_string(),
+                    free_world_bit: country.free_world,
+                }),
+                None => None,
+            }
+        }
+        None => None,
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::neighborhood::gossip::GossipBuilder;
+    use crate::neighborhood::node_location::{get_node_location, NodeLocation};
+    use crate::neighborhood::node_record::{NodeRecord, NodeRecordMetadata};
+    use crate::sub_lib::node_addr::NodeAddr;
+    use crate::test_utils::neighborhood_test_utils::{
+        db_from_node, make_node_record, pick_country_code_record,
+    };
+    use std::net::{IpAddr, Ipv4Addr};
+
+    #[test]
+    fn test_node_location() {
+        let node_location =
+            get_node_location(Some(IpAddr::V4(Ipv4Addr::new(125, 125, 125, 1)))).unwrap();
+
+        assert_eq!(node_location.country_code, "CN");
+        assert_eq!(node_location.free_world_bit, false);
+    }
+
+    #[test]
+    fn construct_node_record_metadata_with_free_world_bit() {
+        let mut metadata = NodeRecordMetadata::new();
+        metadata.node_location_opt = get_node_location(Some(IpAddr::V4(Ipv4Addr::new(1, 1, 1, 1))));
+        assert_eq!(
+            metadata.node_location_opt.as_ref().unwrap(),
+            &NodeLocation {
+                country_code: "AU".to_string(),
+                free_world_bit: true
+            }
+        );
+        assert_eq!(
+            metadata.node_location_opt.as_ref().unwrap().free_world_bit,
+            true
+        );
+        assert_eq!(
+            metadata.node_location_opt.as_ref().unwrap().country_code,
+            "AU"
+        );
+    }
+
+    #[test]
+    fn construct_node_record_for_test() {
+        let mut node_record = make_node_record(1111, true);
+        let country_record = pick_country_code_record(3333);
+        node_record.metadata.node_location_opt = Some(NodeLocation {
+            country_code: country_record.1.clone(),
+            free_world_bit: country_record.2,
+        });
+        node_record.metadata.node_addr_opt =
+            Some(NodeAddr::new(&country_record.0, &[8000 % 10000]));
+        node_record.inner.country_code = country_record.1;
+
+        assert_eq!(
+            node_record.metadata.node_location_opt,
+            Some(NodeLocation {
+                country_code: "AU".to_string(),
+                free_world_bit: true
+            })
+        )
+    }
+
+    #[test]
+    fn create_gossip_node_with_addr_and_country_code_reveal_results_in_node_with_addr_adn_free_world_bit(
```

Comment:
the name is clumsy, and I don't really know how it reads... but maybe it could be improved as:

`node_record_from_gossip_with_addr_and_country_is_populated_with_right_addr_and_free_world_bit`